### ISSUE TITLE
[Feature Request]: Semantic Duplication Check for potential duplicated generated in `generate_synthetic_queries_over_documents`

### ISSUE THREAD

**Issue by AftabHussain** [run-llama/llama_index#20809]
### Feature Description

# Semantic Dedupulication

## Current State: Absence of any Deduplication step

In `llama_index/finetuning/cross_encoders/dataset_gen.py`, `generate_synthetic_queries_over_documents(...)` generates synthetic questions per document chunk using an LLM and aggregates them into a flat list. However, there is currently no deduplication step before returning the final list of questions. Currently, we have (See [full code](https://github.com/run-llama/llama_index/blob/f6b8a6d00d20f222975ac60f3111f0d29d41b462/llama-index-finetuning/llama_index/finetuning/cross_encoders/dataset_gen.py#L35C1-L88C21)):

```py
questions.extend(response_questions)
...
return questions
```
All generated questions are appended and returned as-is, without normalization or deduplication.

This can result in identical questions across multiple chunks, and near-duplicate questions with minor formatting differences. Further down the training pipeline, this can entail redundant training examples in downstream cross-encoder datasets, and inflated dataset size without additional semantic coverage.

This issue becomes more noticeable when `num_questions_per_chunk is large`, and documents contain repetitive structure.

## Proposed Improvement 1: Lightweight deduplication

Before returning questions, do a check for exact matches, preserving order, stripping whitespace, and removing empty strings:

```py
deduped_questions = list(dict.fromkeys(q.strip() for q in questions if q.strip()))
return deduped_questions
```

## Proposed Improvement 2: Semantic Deduplication

**Main goal.** We want near-duplicate detection (paraphrases, trivial rewrites, casing/punctuation changes, “What is X?” vs “Explain X.”)

We may use embedding-based near-duplicate removal -- encode each generated question into a vector embedding and compare it against embeddings of previously accepted questions. If the cosine similarity between a new question and any existing one exceeds a predefined threshold, the new question is treated as a semantic duplicate and discarded. 

To implement, we should first normalize the text (e.g., stripping whitespace, removing numbering or bullet markers, optionally lowercasing), then generate embeddings using a sentence-level model such as OpenAI embeddings or a local sentence-transformers model. 

To ensure scalability, an Approximate Nearest Neighbor (ANN) index (e.g., FAISS or hnswlib) can be used to efficiently retrieve the most similar existing vectors rather than comparing against all prior questions.

**Key tradeoffs to consider.** This semantic deduplication approach requires computing and storing embeddings, as well as maintaining a vector index for similarity search.

### Value of Feature

Deduplication helps further down the training pipeline as follows:

- Cross-encoder training quality improves with diverse queries
- Prevents redundant positives in ranking datasets
- Reduces training time and memory usage
- Encourages better semantic coverage across chunks

**Comment by Nagavenkatasai7** [comment]
Great feature request — this is a real pain point when building synthetic evaluation datasets for RAG pipelines. I've hit this exact issue when generating query sets for cross-encoder fine-tuning at scale.

A few thoughts from practical experience:

**+1 on the two-tier approach.** Lightweight dedup (Improvement 1) should definitely land first as a quick win. In my experience, exact-match + whitespace normalization catches ~15-20% of duplicates with near-zero overhead. The `dict.fromkeys()` pattern you proposed is clean and order-preserving.

**On the semantic dedup threshold (Improvement 2)** — from experimenting with embedding-based dedup in similar pipelines, I'd suggest a cosine similarity threshold around **0.92-0.95** rather than the typical 0.8-0.85 used for retrieval. The reason: for training data, you want to keep questions that are semantically *related* but structurally different (e.g., "What causes X?" vs "Why does X happen?") since they teach the model different query patterns. Too aggressive a threshold kills useful diversity.

**One additional consideration: cluster-aware deduplication.** Instead of pure pairwise dedup, consider grouping questions by their source chunk ID and running dedup *within* each cluster first, then across clusters. This preserves cross-chunk coverage (which is the whole point of generating synthetic queries) while still removing redundancy within each chunk's generated set.

**Practical implementation suggestion:**

```python
from sentence_transformers import SentenceTransformer
import numpy as np

def semantic_dedup(questions, threshold=0.93, model_name="all-MiniLM-L6-v2"):
    model = SentenceTransformer(model_name)
    embeddings = model.encode(questions, normalize_embeddings=True)
    
    keep = []
    for i, emb in enumerate(embeddings):
        if not keep or max(np.dot(emb, embeddings[j]) for j in keep) < threshold:
            keep.append(i)
    
    return [questions[i] for i in keep]
```

This is O(n*k) where k = kept questions, which is much cheaper than all-pairs comparison for large datasets. For production scale, swapping in FAISS as you mentioned would make this sublinear.

Happy to help with a PR if there's interest in moving this forward.
